aiosqlite/core.py

```diff
@@ -19,6 +20,38 @@
 LOG = logging.getLogger('aiosqlite')
 
 
+class _ContextManager(Coroutine):
+    __slots__ = ('_coro', '_obj')
+
+    def __init__(self, coro):
+        self._coro = coro
+        self._obj = None
+
+    def send(self, value):
+        return self._coro.send(value)
+
+    def throw(self, typ, val=None, tb=None):
+        if val is None:
+            return self._coro.throw(typ)
+        if tb is None:
+            return self._coro.throw(typ, val)
+        return self._coro.throw(typ, val, tb)
+
+    def close(self):
+        return self._coro.close()
+
+    def __await__(self):
+        return self._coro.__await__()
+
+    async def __aenter__(self):
+        self._obj = await self._coro
+        return self._obj
+
+    async def __aexit__(self, exc_type, exc, tb):
+        await self._obj.close()
+        self._obj = None
```

Comment:
What is `_ContextManager` providing that we don't get by having `__aenter__/__aexit__` on Cursor directly?  Especially since it's only used when asking for a cursor directly via `db.cursor()` and not on something like `db.execute()`?

Reply:
`_ContextManager` automatically awaits the value on `__aenter__,` so that we can write:
```
async with con.cursor() as cur:
```
instead of
```
async with await con.cursor() as cur:
```

Reply:
OK, can you make that work for more than just `db.cursor()`?  Eg, in your new unit test, you needed to do `async with await db.execute() as cursor`, which is not consistent with `async with db.cursor()`.  

Also, rather than having `Connection.cursor()` chain to `self._cursor()`, just fold the connection manager into the original `cursor()` method, and follow that pattern for any method that returns a cursor.

Reply:
> Also, rather than having Connection.cursor() chain to self._cursor(), just fold the connection manager into the original cursor() method, and follow that pattern for any method that returns a cursor.

It's impossible - the original cursor method is async, and it has to be sync to be able to return ContextManager, and not a coroutine.

Reply:
Hmm, I'd really like to have a consistent interface for all the methods without having duplicated function names to support both use cases.  How about a decorator to apply ContextManager to the results the coroutines?

Eg:

    @contextmanager
    async def cursor(...):
        ...

Reply:
This seems possible, I'll take a look at it tommorow.